index.js

```diff
@@ -3044,81 +3044,151 @@
     impl: at
   };
 
-  //# head :: Array a -> Maybe a
+  //# head :: Foldable f => f a -> Maybe a
   //.
-  //. Returns Just the first element of the given array if the array contains
-  //. at least one element; Nothing otherwise.
+  //. Returns Just the first element of the given structure if it
+  //. contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.head ([1, 2, 3])
   //. Just (1)
   //.
   //. > S.head ([])
   //. Nothing
+  //.
+  //. > S.head (Cons ('foo') (Cons ('bar') (Nil)))
+  //. S.Just ('foo')
+  //.
+  //. > S.head (Nil)
+  //. S.Nothing
   //. ```
+
+  function head(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return maybe.isJust ? maybe : Just (x);
+      },
+      Nothing,
+      foldable
+    );
+  }
   _.head = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
-    impl: array (Nothing) (B (K) (Just))
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
+    impl: head
+
   };
 
-  //# last :: Array a -> Maybe a
+  //# last :: Foldable f => f a -> Maybe a
   //.
-  //. Returns Just the last element of the given array if the array contains
-  //. at least one element; Nothing otherwise.
+  //. Returns Just the last element of the given structure if it
+  //. contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.last ([1, 2, 3])
   //. Just (3)
   //.
   //. > S.last ([])
   //. Nothing
+  //.
+  //. > S.last (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil))))
+  //. S.Just ('baz')
+  //.
+  //. > S.last (Nil)
+  //. S.Nothing
   //. ```
-  function last(xs) {
-    return xs.length > 0 ? Just (xs[xs.length - 1]) : Nothing;
+
+  function last(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return Just (x);
+      },
+      Nothing,
+      foldable
+     );
   }
+
   _.last = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
     impl: last
   };
 
-  //# tail :: Array a -> Maybe (Array a)
+  //# tail :: (Applicative f, Foldable f, Monoid f) => f a -> Maybe (f a)
```

Comment:
```suggestion
  //# tail :: (Applicative f, Foldable f, Monoid (f a)) => f a -> Maybe (f a)
```

It's not yet possible to precisely describe this function's type using `def`, but the textual signature can be precise. The third constraint should be `Monoid (f a)` rather than `Monoid f`. `Maybe a`, for example, satisfies Monoid only if `a` satisfies Semigroup ([source][1]). The question we want to ask is not whether `Maybe` satisfies Monoid, but whether `Maybe String` satisfies Monoid. Does this make sense?


[1]: https://github.com/sanctuary-js/sanctuary-maybe/blob/v1.1.0/index.js#L117

Reply:
yeah that makes sense

Reply:
🆗